casts/splitNavigator/src/SplitNavigator/SplitRouter.ts

```diff
@@ -49,246 +50,680 @@ export type SplitActionHelpers = {
+                routes: applyTabNavigateActionToRoutes(state, action, options, mainIndex),
+            };
+        }
+        return null;
+    }
+}
+
+export function SplitRouter(routerOptions: SplitRouterOptions) {
+    // eslint-disable-next-line prefer-const
+    let { isSplitted } = routerOptions;
+    let isInitialized = false;
+    const foldedRouter = new SplitFoldedRouter(routerOptions);
+    const unfoldedRouter = new SplitUnfoldedRouter(routerOptions);
+    const router: Router<SplitNavigationState, CommonNavigationAction | SplitActionType> = {
+        ...BaseRouter,
+
+        // Every router in react-navigation should have a type
+        // and it should be consistent between re-renders
+        // or library will try to re-initialize the state
+        // every time
+        type: 'split',
+
+        getInitialState(options: RouterConfigOptions) {
+            return (isSplitted ? unfoldedRouter : foldedRouter).getInitialState(options);
         },
 
-        getRehydratedState(state, params): SplitNavigationState {
+        /**
+         * Usually it's called for linking, as the result of it is a partial state
+         * i.e. for /foo/bar it's going to be sth like this
+         * {
+         *   stale: true,
+         *   routes: [{
+         *     name: 'foo',
+         *     state: {
+         *       stale: true,
+         *       routes: [{
+         *         name: 'bar'
+         *       }]
+         *     }
+         *   }]
+         * }
+         */
+        getRehydratedState(
+            state: SplitNavigationState | PartialState<SplitNavigationState>,
+            options,
+        ): SplitNavigationState {
             const isStale = state.stale;
-            let newState: SplitNavigationState;
 
             if (isStale === false) {
                 return state as SplitNavigationState;
             }
 
-            if (isSplitted) {
-                newState = tabRouter.getRehydratedState(state as any, params) as any;
-                newState = this.ensureTabState(newState);
-            } else {
-                newState = stackRouter.getRehydratedState(state as any, params) as any;
-                newState = this.ensureStackState(newState);
-            }
-
-            Object.assign(newState, { type: router.type, isSplitted });
-            return newState;
+            return (isSplitted ? unfoldedRouter : foldedRouter).getRehydratedState(state, options);
         },
 
-        getStateForRouteNamesChange(state, options) {
-            const newState: SplitNavigationState = isSplitted
-                ? (tabRouter.getStateForRouteNamesChange(state as any, options) as any)
-                : stackRouter.getStateForRouteNamesChange(state as any, options);
-
-            Object.assign(newState, { type: router.type, isSplitted });
-            return newState;
+        getStateForRouteNamesChange(state /* , options */) {
+            console.warn("Dynamic routes isn't supported yet in SplitRouter");
```

Comment:
Should it be supported?)

Reply:
For now, I guess not, as we don't have dynamic routes in Surf